### DISCUSSION TITLE
Syntatic sugar ( macro ) for inheritAttrs and name.

### DISCUSSION THREAD

**Discussion by fenilli** [vuejs/rfcs#431] [RFC Discussions]
# Summary
This proposal is close related to #430, but aims to be constricted to these attributes.
Introduces macros in script setup, specifically to be able to set inheritAttrs and name.

# Motivation
Writing components that should pass down props and attrs to the correct tags in the new script setup is verbose, and tedious. So this proposal tries to fix this issue by adding macros to merge or overwrite those attributes.

## Example
```vue
<script setup lang="ts">
$componentName('My Component');
$inheritAttrs(false);
</script>
```
<details>
  <summary>Equivalent Output</summary>
  
  ```vue
  <script lang="ts">
  export default {
    name: 'My Component',
    inheritAttrs: false,
  }
  </script>

  <script setup lang="ts">
  </script>
  ```
</details>

## Example with user defined script tag
```vue
<script lang="ts">
export default {
  inheritAttrs: true,
}
</script>
  
<script setup lang="ts">
$componentName('My Component');
$inheritAttrs(false);
</script>
```
<details>
  <summary>Equivalent Output</summary>
  
  ```vue
  <script lang="ts">
  export default {
    name: 'My Component',
    inheritAttrs: false,
  }
  </script>

  <script setup lang="ts">
  </script>
  ```
</details>

# Behaviour
1. The macros should create a <script> tag if there isn't one already defined.
2. If <script> tag is defined by the user it should merge or overwrite their respective values.
3. Macro definition should always be prefered when overwriting is needed.

**Comment by sxzz** [comment]
IMHO, this proposal seems to be too restricted. `defineOptions` can be used in [this scenario](https://twitter.com/enzoinnocenzi/status/1494177242074828803).

**Reply by HerringtonDarkholme** [reply]
Indeed, to support custom option field defined in user libraries, this RFC will need to provide new compiler macros. e.g. the linked example users will need `$layout(wide)`.

**Reply by fenilli** [reply]
isn't that the problem with the defineOptions RFC? that is not restrictive enough, because you can't really use defineProps and defineOptions({props}) together, and that supporting custom option fields defined by other libraries aren't supposed to be inside the vue's macro?

**Comment by fenilli** [comment]
The proposal could be changed to be `useMisc` instead, where only miscellaneous properties should be passed down.

```vue
<script lang="ts">
export default {
  inheritAttrs: true,
}
</script>
  
<script setup lang="ts">
useMisc({
  inheritAttrs: false,
});
</script>
```
